File: data_structures/trees/comparing_tree_structure/python/comparing_tree_structure.py
class Node:
    def __init__(self, val, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

def compare(a, b):
    if a is None and b is None:
        return True
    
    if a is None or b is None:
        return False
    
    if a.val != b.val:
        return False
    
    return compare(a.left, b.left) and compare(a.right, b.right)


if __name__ == "__main__":
    """Comparing one tree to another ensuring structure and contents are identical"""
    tree_one = Node(4, Node(2, None, None), Node(5, None, None))
    tree_two = Node(4, Node(2, None, None), Node(5, None, None))
    tree_three = Node(4, Node(2, Node(5, None, None)))

    print(compare(tree_one, tree_two))

    print(compare(tree_one, tree_three))

File: data_structures/trees/comparing_tree_structure/java/comparingTreeStructure.java
class comparingTreeStructure
{
    public static boolean compare(Node a, Node b)
    {
        if(a == null && b == null)
        {
            return true;
        }
        else if(a == null || b == null)
        {
            return false;
        }
        else if(a.val != b.val)
        {
            return false;
        }

        return compare(a.left, b.left) && compare(a.right, b.right);

    }

    public static void main(String[] args) {
        //Comparing one tree to another ensuring structure and contents are identical

        Node th_one = new Node(4);
        th_one.left = new Node(2);
        th_one.right = new Node(5);

        Node th_two = new Node(4);
        th_two.left = new Node(2);
        th_two.right = new Node(5);

        Node th_three = new Node(4);
        th_three.left = new Node(2);
        th_three.left.right = new Node(5);

        System.out.println(compare(th_one, th_two));

        System.out.println(compare(th_one, th_three));
    }
}